# CPSC 483-01: Project 3, Group 5, Fall 2021 
**Group Members:**
- Nicole Serna
- Zachary Serna

In [1]:
#Experiment 1
import pandas as pd
#Complete dataSet
dataSet = pd.read_csv('bank-additional-full.csv', delimiter=";")
df = pd.DataFrame(dataSet)
df.head(n=5)

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

Experiment 1 Results/Comments

We use pandas to load the csv file in, examining the results.

In [2]:
#Experiment 2

#import both sklearn and numpy
import numpy as np
import sklearn
from sklearn.model_selection import train_test_split

#split the data 90/10 into training and test sets respectively
X_train, X_test, y_train, y_test= sklearn.model_selection.train_test_split(dataSet, df.loc[:,"y"],test_size=.1,train_size=.9, random_state=(2021-10-25))

Experiment 2 Results/Comments

We use train_test_split to split the set 90/10 and assign the sets to their appropriate variables in the process.

In [3]:
#Experiment 3

#We drop feature y from the set after saving it to variable 'y'
y = pd.DataFrame(dataSet, columns=["y"])
features = df.drop(columns=["duration", "y"])
X_train, X_test, y_train, y_test= sklearn.model_selection.train_test_split(features, y,test_size=.1,train_size=.9, random_state=(2021-10-25))

Experiment 3 Results/Comments

After assigning the y feature to its respective variable, we drop it from the rest of the features.

In [4]:
#Experiment 4


bank_client_data_train = pd.DataFrame(X_train, columns=["age","job","marital","education","default","housing","loan"])
bank_client_data_test = pd.DataFrame(X_test, columns=["age","job","marital","education","default","housing","loan"])

#Properly set the data for future use using get_dummies
dummy_bcd_train = pd.get_dummies(bank_client_data_train,drop_first=True)
dummy_bcd_test = pd.get_dummies(bank_client_data_test,drop_first=True)

Experiment 4 Results/Comments

Since the data we are working with is categorical, we use get dummies to turn our data into quantitative data that we can analyze.

In [5]:
#Experiment 5

#Import Categorical NB, then fit it to the bcd_training set and y_train
from sklearn.naive_bayes import CategoricalNB
cnb = CategoricalNB()
cnb.fit(dummy_bcd_train,y_train)

#Output result
print(cnb.score(dummy_bcd_test,y_test))
print(cnb.score(dummy_bcd_train,y_train))

C:\Users\nikki\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


0.8788540907987376
0.8818689470986538


Experiment 5 Results/Comments

training score:    0.8818689470986538

test score:        0.8788540907987376

Overall the classifier was fairly accurate. 88% is a fairly high score showing CNB fits the data well.

In [6]:
#Taking a look at the data used in the previous experiment (expirement 5)
print(dummy_bcd_train)

       age  job_blue-collar  job_entrepreneur  job_housemaid  job_management  \
22975   56                0                 0              0               0   
14746   26                0                 0              0               0   
12505   38                0                 0              0               0   
6801    33                0                 0              0               0   
18389   29                0                 0              0               0   
...    ...              ...               ...            ...             ...   
17987   31                1                 0              0               0   
20299   29                0                 0              0               0   
14556   49                1                 0              0               0   
28981   47                0                 0              0               0   
32764   35                0                 0              0               0   

       job_retired  job_self-employed  

Experiment 6 Response

Even though Categorical Naive Bayes assumes that each value of the age variable is a separate category, in our data used in the previous expeirment, there is only one category for the age variable. This is not reasonable when using the Categorical Naive Bayes classifier, since it is assuming that each age is a serparete column.

In [7]:
#Experiment 7


#Set up for division
bins = [0,10,20,30,40,50,60,70,80,90,100]
labels = ['0-10','11-20','21-30','31-40','41-50','51-60','61-70','71-80','81-90','91-100']
bcd_train_ageSplit = bank_client_data_train

#splits age into appropriate bins and then use get_dummies to set the data up
#for later use
bcd_train_ageSplit['Age_Category'] = pd.cut(bank_client_data_train['age'], bins= bins, labels= labels)
bcd_train_ageSplit = bcd_train_ageSplit.drop(columns=['age'])
dummy_bcd_train_ageSplit = pd.get_dummies(bcd_train_ageSplit,drop_first=True)

bcd_test_ageSplit = bank_client_data_test
bcd_test_ageSplit['Age_Category'] = pd.cut(bank_client_data_test['age'], bins= bins, labels= labels)
bcd_test_ageSplit = bcd_test_ageSplit.drop(columns=['age'])
dummy_bcd_test_ageSplit = pd.get_dummies(bcd_test_ageSplit,drop_first=True)

#fits data to see how well it fits the CNB classifier
cnb.fit(dummy_bcd_train_ageSplit,y_train)
print(cnb.score(dummy_bcd_test_ageSplit,y_test))
print(cnb.score(dummy_bcd_train_ageSplit,y_train))

C:\Users\nikki\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


0.8788540907987376
0.881437319593191


Experiment 7 Results/Comments

training score: 0.881437319593191

test score:     0.8788540907987376

Based on the results above, we can see setting the data into bins had little to no effect on the data, only VERY slightly decreasing the accuracy of the test set.

In [8]:
#Experiment 8


#KNN classifier, we chose to use the default number of neighbors: 5
from sklearn import neighbors
knn = neighbors.KNeighborsClassifier(n_neighbors=5)
knn.fit(dummy_bcd_train,y_train)
print(knn.score(dummy_bcd_test,y_test))
print(knn.score(dummy_bcd_train,y_train))

C:\Users\nikki\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:179: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


0.8771546491866958
0.8903935903315439


Experiment 8 Results/Comments

KNN training score:    0.8903935903315439

KNN test score:        0.8771546491866958


Based on the results above, we can see that both the test and training scores are extremely similar to those of the CategoricalNB classifier.

In [9]:
#Experiment 9

#returns counts of no(0) and yes(1)
y_test['y'].value_counts()

no     3646
yes     473
Name: y, dtype: int64

Experiment 9 Results/Comments

no     3646

yes     473p

Name: y, dtype: int64

We can see from the results above that 3646 have response 0 (no), while 473 have response 1 (yes). If we simply assumed that no customer ever subscribed to the product, then the score would be 3646/4119, which would be roughly 88%.

In [10]:
#Experiment 10

#we import the confusion matrix and roc funcs from sklearn
from sklearn.metrics import confusion_matrix
from sklearn.metrics import roc_auc_score
from sklearn.metrics import roc_curve
from sklearn.linear_model import LogisticRegression

#create a dummy set of only 0's and score it against our y_test
dumb_class = np.zeros_like(y_test, dtype='uint')
y_test_dummy = pd.get_dummies(y_test, drop_first =True)
conf_matrix = confusion_matrix(y_test_dummy.values,dumb_class)
print("Confusion Matrix:\n", conf_matrix)

#Calculating AUC score 
score_auc = roc_auc_score(y_test_dummy.values,dumb_class)
print("AUC Score:\n", score_auc )

Confusion Matrix:
 [[3646    0]
 [ 473    0]]
AUC Score:
 0.5


Experiment 10 Results/Comments

Confusion Matrix:

 [[3646    0]
 
 [ 473    0]]
 
AUC Score:
 0.5


We end up with a AUC score of .5 which is much lower than our previous results

In [11]:
#Experiment 11

#We use the predict and score functions on our ageSplit set
#we then use the result of predict for our confusion matrix
pred_bins = cnb.predict(dummy_bcd_test_ageSplit)
score_bins = cnb.score(dummy_bcd_test_ageSplit,y_test)
print("Score:", score_bins)
conf_matrix_bins = confusion_matrix(pred_bins, y_test)
print("Confusion Matrix:\n", conf_matrix_bins)


#Calculating AUC for ageSplit set
predictProb_y = cnb.predict_proba(dummy_bcd_test_ageSplit)[:,1]
fps, tps, thresholds = roc_curve(y_test, predictProb_y,pos_label=1)
roc_score = roc_auc_score(y_test, predictProb_y)
print("CategoricalNB - ROC AUC score", roc_score)

#Confusion Matrix for KNN set
pred_KNN = knn.predict(dummy_bcd_test)
conf_matrix_KNN = confusion_matrix(pred_KNN, y_test)
print("Confusion Matrix KNN:\n", conf_matrix_KNN)

#Calculating AUC for KNN
predictProb_KNN = knn.predict_proba(dummy_bcd_test)[:,1]
roc_score_KNN = roc_auc_score(y_test, predictProb_KNN)
print("KNN - ROC AUC score", roc_score_KNN)


Score: 0.8788540907987376
Confusion Matrix:
 [[3554  407]
 [  92   66]]
CategoricalNB - ROC AUC score 0.6389724207594062


C:\Users\nikki\anaconda3\lib\site-packages\sklearn\metrics\_ranking.py:949: UndefinedMetricWarning: No positive samples in y_true, true positive value should be meaningless
  warnings.warn("No positive samples in y_true, "


Confusion Matrix KNN:
 [[3559  419]
 [  87   54]]
KNN - ROC AUC score 0.6142052050438431


Experiment 11 Results/Comments

Confusion Matrix:

 [[3554  407]
 
 [  92   66]]


CategoricalNB - ROC AUC score 0.6389724207594062

Confusion Matrix KNN:

 [[3554  407]
 
 [  92   66]]
 
KNN - ROC AUC score 0.6142052050438431



The classifiers are performing much more poorly than shown in our previous experiments, showing a potential imbalance in the data.

In [12]:
#Experiment 12

y_0_count = (dataSet.y == 'no').sum()
y_no_df = dataSet.drop(dataSet[dataSet.y == 'yes'].index)
dataSetWeight = dataSet
dataSetWeight['weight'] = 1
dataSetWeight['weight'] = dataSetWeight['weight'].where(dataSet.y == 'yes', '0')
y_overSample = dataSetWeight.sample(n = y_0_count, weights = 'weight', replace = True, random_state=(2021-10-25))
df_overSample = pd.concat([y_no_df, y_overSample], axis=0)

#Recreating test/training set, seperating target variable
y_OS = pd.DataFrame(df_overSample, columns=["y"])
features_OS = df_overSample.drop(columns=["duration", "y", "weight"])
X_train_OS, X_test_OS, y_train_OS, y_test_OS= sklearn.model_selection.train_test_split(features_OS, y_OS,test_size=.1,train_size=.9, random_state=(2021-10-25))

#dropping the non-bank client data columns
#running get dummies again for categorical variables
bcd_OS_train = pd.DataFrame(X_train_OS, columns=["age","job","marital","education","default","housing","loan"])
bcd_OS_test = pd.DataFrame(X_test_OS, columns=["age","job","marital","education","default","housing","loan"])
dummy_bcdOS_train = pd.get_dummies(bcd_OS_train,drop_first=True)
dummy_bcdOS_test = pd.get_dummies(bcd_OS_test,drop_first=True)
#because of the way our train/test sets are split, the test set ends up with no values for this column, so it deletes it
#but in order to run CNB we need the train and test set to have the same number of columns
#so we add the 'default_yes' column to the test set but fill it with all zeros
dummy_bcdOS_test['default_yes'] = 0

Experiment 12 Results/Comments


We create a new dataset which places more weight on response 1, which would be 'yes' in this scenario. To do this we oversample users with response yes, in order to have an even split of individuals with response yes and individuals with response no.

In [13]:
#Experiment 13

#retraining categorical naive bayes
cnb = CategoricalNB()
cnb.fit(dummy_bcdOS_train,y_train_OS)
print(cnb.score(dummy_bcdOS_test,y_test_OS))
print(cnb.score(dummy_bcdOS_train,y_train_OS))

pred_OS_cnb = cnb.predict(dummy_bcdOS_test)
conf_matrix_cnb_OS = confusion_matrix(pred_OS_cnb, y_test_OS)
print("Confusion Matrix CategoricalNB:\n", conf_matrix_cnb_OS)

predictProb_y_OS_cnb = cnb.predict_proba(dummy_bcdOS_test)[:,1]
roc_score_cnbOS = roc_auc_score(y_test_OS, predictProb_y_OS_cnb)
print("CategoricalNB - ROC AUC score", roc_score_cnbOS)

#retrain KNN with balanced data
knn = neighbors.KNeighborsClassifier(n_neighbors=5)
knn.fit(dummy_bcdOS_train,y_train_OS)
##print(knn.score(dummy_bcdOS_test,y_test_OS))
##print(knn.score(dummy_bcdOS_train,y_train_OS))
pred_KNN_OS = knn.predict(dummy_bcdOS_test)
conf_matrix_KNN_OS = confusion_matrix(pred_KNN_OS, y_test_OS)
print("Confusion Matrix KNN:\n", conf_matrix_KNN_OS)

predictProb_KNN_OS = knn.predict_proba(dummy_bcdOS_test)[:,1]
roc_score_KNN_OS = roc_auc_score(y_test_OS, predictProb_KNN_OS)
print("KNN - ROC AUC score", roc_score_KNN_OS)

C:\Users\nikki\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


0.6049247606019151
0.609354573921503
Confusion Matrix CategoricalNB:
 [[2325 1560]
 [1328 2097]]
CategoricalNB - ROC AUC score 0.6541865230992601


C:\Users\nikki\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:179: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


Confusion Matrix KNN:
 [[2361 1337]
 [1292 2320]]
KNN - ROC AUC score 0.6917650252963896


Experiment 13 Results/Comments

Confusion Matrix CategoricalNB:
 [[2325 1560]
 
 [1328 2097]]
 
CategoricalNB - ROC AUC score 0.6541865230992601

Confusion Matrix KNN:
 [[2361 1337]
 
 [1292 2320]]
KNN - ROC AUC score 0.6917650252963896


As we can see from the results above, The KNN classifier performs slightly better than the CNB classifier.

In [14]:
#Experiment 14
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()

bc_data_train = pd.DataFrame(X_train, columns=[ "emp.var.rate", "cons.price.idx", "cons.price.idx", "cons.conf.idx", "euribor3m" "nr.employed"])
bc_data_test = pd.DataFrame(X_test, columns=[ "emp.var.rate", "cons.price.idx", "cons.price.idx", "cons.conf.idx", "euribor3m" "nr.employed"])

bc_data_train = bc_data_train.fillna(0)
bc_data_test = bc_data_test.fillna(0)


#Properly set the data for future use using get_dummies
#dum_bcd_train = pd.get_dummies(bc_data_train,drop_first=True)
#dum_bcd_test = pd.get_dummies(bc_data_test,drop_first=True)



#dum_bcd_train.fillnaN(0)
#Import Categorical NB, then fit it to the bcd_training set and y_train
gnb.fit(bc_data_train,y_train)

gnb.predict(bc_data_test)

#Output result
print(gnb.score(bc_data_test,y_test))
print(gnb.score(bc_data_train,y_train))

#Confusion Matrix
pred_gnb = gnb.predict(bc_data_test)
conf_matrix_gnb = confusion_matrix(pred_gnb, y_test)
print("Confusion Matrix CategoricalNB:\n", conf_matrix_gnb)

#AUC score
predictProb_gnb = gnb.predict_proba(bc_data_test)[:,1]
roc_score_gnb = roc_auc_score(y_test, predictProb_gnb)
print("CategoricalNB - ROC AUC score", roc_score_gnb)


0.8720563243505706


C:\Users\nikki\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


0.8761768593703634
Confusion Matrix CategoricalNB:
 [[3490  371]
 [ 156  102]]
CategoricalNB - ROC AUC score 0.7016899402629543


Experiment 14 Results/Comments

training score:     0.8761768593703634

test score:         0.8720563243505706

Confusion Matrix CategoricalNB:
 [[3490  371]
 [ 156  102]]
CategoricalNB - ROC AUC score 0.7016899402629543

Overall, the data does a subpar job of predicting the data. While it is an improvement over previous AUC scores we have observed, 70% is still a relatively low number in terms of reliability.

In [15]:
#Experiment 15


#Pull columns/features we are interested in analyzing
sac_data_train = pd.DataFrame(X_train_OS, columns=[ "emp.var.rate", "cons.price.idx", "cons.price.idx", "cons.conf.idx", "euribor3m" "nr.employed"])
sac_data_test = pd.DataFrame(X_test_OS, columns=[ "emp.var.rate", "cons.price.idx", "cons.price.idx", "cons.conf.idx", "euribor3m" "nr.employed"])

sac_data_train = sac_data_train.fillna(0)
sac_data_test = sac_data_test.fillna(0)

#Import Categorical NB, then fit it to the bcd_training set and y_train
gnb.fit(sac_data_train,y_train_OS)

gnb.predict(sac_data_test)

#Output result
print(gnb.score(sac_data_test,y_test_OS))
print(gnb.score(sac_data_train,y_train_OS))

#Confusion Matrix
#print
pred_gnb_balance = gnb.predict(sac_data_test)
conf_matrix_gnb_balance = confusion_matrix(pred_gnb_balance, y_test_OS)
print("Confusion Matrix CategoricalNB:\n", conf_matrix_gnb_balance)

predictProb_gnb_balance = gnb.predict_proba(sac_data_test)[:,1]
roc_score_gnb_balance = roc_auc_score(y_test_OS, predictProb_gnb_balance)
print("CategoricalNB - ROC AUC score", roc_score_gnb_balance)

C:\Users\nikki\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


0.6574555403556771
0.6518104155899431
Confusion Matrix CategoricalNB:
 [[2710 1561]
 [ 943 2096]]
CategoricalNB - ROC AUC score 0.7111010230465242


Experiment 15 Results/Comments

Training Score: 0.6518104155899431

Test Score:     0.6574555403556771

Confusion Matrix CategoricalNB:

 [[2710 1561]
 
 [ 943 2096]]
 
CategoricalNB - ROC AUC score 0.7111010230465242


The results do change. Both the training and test score decrease; however, the AUC score increases very slightly.